src/main/java/net/fabricmc/loader/api/metadata/AbstractModDependency.java

```diff
@@ -0,0 +1,69 @@
+package net.fabricmc.loader.api.metadata;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import net.fabricmc.loader.api.VersionRange;
+
+public abstract class AbstractModDependency implements ModDependency {
+	protected final String modId;
+	private Collection<VersionRange> ranges;
+
+	protected AbstractModDependency(String modId) {
+		this.modId = modId;
+	}
+
+	@Override
+	public String getModId() {
+		return modId;
+	}
+
+	protected abstract String[] getVersionMatchers();
+
+	@Override
+	public Collection<VersionRange> getVersionRanges() {
+		if (ranges == null) {
+			ranges = Stream.of(getVersionMatchers()).map(matcher -> {
+				char firstChar = matcher.charAt(0);
+				char secondChar = 0;
+				if (matcher.length() > 1)
+					secondChar = matcher.charAt(1);
+				switch (firstChar) {
```

Comment:
Also maintainability concern here. Should add to `VersionRange` so we have an easier time maintaining this. Leaving this here will most likely make us forget this code and make updates to version range code out of sync.

Reply:
moved to `static Collection<VersionRange> VersionRange.fromVersionMatchers`